existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,303 @@
+
+First of all, you can now change the order of type variables.
+Let's imagine you want to define a function with 10 type variables (yeah, there are real-life [examples](https://hackage.haskell.org/package/leancheck-0.9.10/docs/Test-LeanCheck-Utils-TypeBinding.html#v:-45--62--62--62--62--62--62--62--62--62--62--62--62-:) of such functions), and you know that when you use it, you want to specify type of the last argument explicitly.
+To pass a type as a parameter to a function, we will use the `TypeApplication` extension.
+Few notes about passing types to functions. Using `TypeApplication` extension you can pass types as function arguments, using `@` sign (If specifying concrete type is unnecessary you can use type wildcard `@_`).
+By default, the order of type variables, which you assign types to, is the same as the order in which they are introduced in the function definition.
+Without `forall`, you would write something like this:
+
+```Haskell
+veryLongFunction :: a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @_ @_ @_ @_ @_ @_ @_ @_ @_ @Integer ...
+```
+Quite long, right?
+However, by using explicit `forall`, this can be simplified:
+
+```Haskell
+veryLongFunction :: forall j a b c d e f g h i. a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @Integer ...
+```
+
+Since `j` is the first type variable in the declaration of `veryLongFunction` you can specify only it.
+This is useful not only in such big examples but also when your function requires any of its arguments to be specified on usage.
+
+### Support of other extensions
+
+`ExplicitForAll` shines most when you need to enable other extensions – some of them just won't work without the `forall` quantifier. :)
+The most common one is `ScopedTypeVariables`, which allows you to pass a type variable from an outer function to the one defined in a where clause.
+
+```haskell
+example :: a -> [a] -> [a]
+example x rest = pair ++ rest
+  where
+    pair :: a -> [a]
+    pair x = [x, x]
+```
+This will cause an error without `ScopedTypeVariables`, because compiler thinks that `a` in `example`'s signature and `a` in `double`'s are two different types.
+To fix it, you also need to use extension and add a `forall` quantifier, as extension won't work without it.
+
+```haskell
+example :: forall a. a -> [a] -> [a]
+```
+You might also notice how it was useful with the `TypeApplication` extension.
+Other extensions that somehow benefit from the usage of `ExplicitForAll` are `LiberalTypeSynonyms`, `RankNTypes`, and many more.
+
+## Existential quantification
+As I said earlier, besides universally quantified variables, Haskell also supports existential types.
+Remember, that we still use the `forall` keyword ¯\\\_(ツ)_/¯.
+Actually, it is implemented this way, because in terms of first-order predicate logic these two constructions are equivalent: `(exists x. p x) -> q` and `forall x. (p x -> q)`.
```

Comment:
Is mentioning this equivalence and predicate logic necessary? I fear it might distract the reader from the point of the blog post.

Connecting the dots between the "predicate logic encoding" and the "haskell encoding" of existential quantification could be useful _if_ we expected most readers to be familiar with predicate logic, but I don't think that's a reasonable expectation...

Reply:
Well, it is good to mention, why something called existential instead is written using `forall`. Even if the reader doesn't know anything about FOPL they will still notice, that there is some clever explanation to it

Reply:
I think it is good to leave it there since it gives some proof that what we are doing with forall has connection with existentials. It's nice to spell it out.